### ISSUE TITLE
[snap] migration from strict to classic enclosure

### ISSUE THREAD

**Issue by shiftkey** [shiftkey/desktop#116]
There's some outstanding problems with our use of the strict enclosure for the Snap installer:
  - #95
  - #98
  - #59

Essentially, a Snap in a strict enclosure is not able to launch any processes outside it sandbox. This affects some of the important integration for Desktop, such as being able to launch your configured editor or shell. 

After talking it over with various Snap team members there appears to be no easy solution for this, other than switching enclosures. But that in itself is a really hard problem, because there's no way to upgrade a package which changes it's enclosure setting from `strict` to `classic` (which is basically unrestricted). This is why we've temporarily disabled the listing on the Snap store to prevent more people installing this app and then facing this problem: #108 

The solution we have figured out is a bit rough, but looks like this (paraphrasing an earlier discussion with @flexiondotorg about this):

### Preparations

 - ✅Make a request for GitHub Desktop to be approved to [use classic confinement](https://forum.snapcraft.io/t/github-desktop-classic-request/10571)
 - ✅Create a branch named [`snap-strict-enclosure`](https://github.com/shiftkey/desktop/compare/linux...shiftkey:snap-strict-enclosure). This will contain the old version with the `strict` confinement that we need to use later.
 - ✅ Create a pull request to restore the `classic` confinement: #117 
 - ✅ Create a pull request to add a prompt for users running `edge` with the `strict` enclosure to manually migrate to the `beta` channel with the `classic` enclosure: #118

### This Week

 - ✅ _Friday 29th March_
   -  [x] Merge #117 into `linux`
   -  [x] bump the version to `release-1.6.5-linux6`
   -  [x] tag and publish from `linux` **to the `beta` channel**
   -  [x] confirm the version in the `beta` channel is working as expected with the classic enclosure  

### Next Week

 - ✅ _Monday 1st April_ -> _Sunday 7th April_
   -  [x] add documentation to README about Snap migration - #135 
   -  [x] Merge #118 into `snap-strict-enclosure` 
   -  [x] bump the version to `release-1.6.5-linux7` so it is ahead of `beta`
   -  [x] tag and publish the snap from `snap-strict-enclosure` **to the `edge` channel**
   - [x] as strict enclosure is unable to invoke `spawnSync` we need a more robust way to detect the versions installed on `edge`
   -  [x] confirm the version in the `edge` channel is working as expected
   -  [x] make the snap public again so new users can install (either from `edge` or `beta`) - closes #108 #95 #98 #59

At this point existing users of the `edge` channel will receive the update and see the notification to transition manually to the `beta` channel.

### Post-Migration

 - [ ] allow time for `edge` installs to migrate themselves over to `beta`
 - [ ] at some point, publish a version on `edge` that uses the classic enclosure
 - [ ] cleanup the Snap migration warning once a sufficient % of users have migrated

**Comment by shiftkey** [comment]
> * Make a request for GitHub Desktop to be approved to use classic confinement (I believe this is underway)

@flexiondotorg is there anything I can do to move this along? All the subsequent work is not possible currently...

**Comment by shiftkey** [comment]
I've refreshed the issue body to reflect the current state of affairs, and hopefully on Friday I can start the migration process.

**Comment by flexiondotorg** [comment]
@shiftkey Looks like you've got everything covered in the sequence of events.

**Comment by shiftkey** [comment]
This is on hold at the moment because the package gets rejected on publish with this message:

```
Error while processing...
The store was unable to accept this snap.
   - confinement 'classic' not allowed with plugs/slots
```

This is the new config we are using to build the snap, and I'm not sure at a glance what I should be changing in here:

https://github.com/shiftkey/desktop/blob/51694bf834ac5c61e90c4e18540c613fb64f5326/script/electron-builder-linux.yml#L40-L48

I'll see if I can troubleshoot more over the weekend...

EDIT: should be addressed with https://github.com/shiftkey/desktop/commit/2af83f572ce16580eb343e5bd1ce61fc8775bff7

EDIT 2: ugh, now it's a different error

```
Error while processing...
The store was unable to accept this snap.
   - confinement 'classic' not allowed with plugs/slots
   - invalid plugs entry (empty)
```

EDIT 3: I've isolated the issue to https://github.com/electron-userland/electron-builder/issues/2730 which linked off to a workaround about manually editing `snap.yaml` and repackaging the app, which was done in 0ca1184fa1602ca5feda164f6e0131d0f4fbb5ce. Then I needed to refresh my access token to support `beta` publishing, so hopefully `-linux6` is the lucky one.

EDIT 4: we have a working beta version, aiming to do some testing this week and confirm it's working as expected..

**Comment by shiftkey** [comment]
Did a quick test pass of the first run and integrations with the latest `beta` and everything looks fine. I opened #134 to track the protocol handler integration, as the earlier work in #110 to support this with the `.deb` package is not relevant here.

**Comment by shiftkey** [comment]
Pausing this because my earlier idea to "just spawn `snap`" here worked in development but not in the strict enclosure after publishing to `edge`:

https://github.com/shiftkey/desktop/blob/320181c419691d9d2c9b8cbf2ad4300e5b97545f/app/src/ui/snap-migration-guide/detect-snap-install.ts#L10-L15

I have an idea to read `process.execPath` which will give me:

 - whether this is a published snap
 - the version embedded in the path

<img width="1125" src="https://user-images.githubusercontent.com/359239/55328176-24a78f80-5462-11e9-9b8a-21e27c3bfe7a.png">

I can whitelist versions that we've published to `beta` to ignore part of this check, which can be obtained by running ` snapcraft revisions --arch amd64 github-desktop`. I don't plan to publish the same build to multiple channels during this transition period, so I think this is fine.

**Comment by shiftkey** [comment]
After poking at available information for `beta` and `edge` channels I noticed a predictable differentiator between the two environments: in the `strict` enclosure the `--disable-gpu-compositing` flag is passed into the renderer process:

<img width="1131" src="https://user-images.githubusercontent.com/359239/55652354-a0d30780-57c1-11e9-8bf8-e015b6e8e8c7.png">

I haven't been able to find the relevant part of the toolchain that's responsible for this, but it seems a pretty reliable indicator of the environment, rather than hard-coding versions which aren't known until they're published to the store.

EDIT: this was a red herring, and not something controlled by the channel - likely Electron itself detecting GPU compositing was disabled at the time (power saving)? Oh well, the `snap-strict-enclosure` branch is not intending to be merged back into the main branch here, so the original naive patch can remain.

**Comment by shiftkey** [comment]
I've tested out the `edge` upgrade warning the user and asking the user to change to the `beta` channel, so I've now made the snap public again and updated the related issues with notes about the migration process.

I'm going to let this issue sit a bit and monitor for feedback.

**Comment by flexiondotorg** [comment]
Thanks for working on this. I'll be installing it when I get home 👍😀

**Comment by shiftkey** [comment]
It's been almost 2 months since I did the cutover to beta and published the dialog about getting people to migrate from `edge` to `beta`. We had almost 400 installs at the time:

<img width="269" alt="Screen Shot 2019-05-30 at 3 39 39 PM" src="https://user-images.githubusercontent.com/359239/58655746-33d77c00-82f1-11e9-9c41-6b941771fbb6.png">

Fast-forward to today and we still have 300+ active installs of `edge`:

<img width="439"   src="https://user-images.githubusercontent.com/359239/58655745-33d77c00-82f1-11e9-9be7-bcd794299012.png">

I was really hoping that this number would be much lower, as I'd love to start exploring using `edge` as a preview channel for things like Electron upgrades and experimentation. But I can't do that if users are ignoring the warning and still running the wrong enclosure.

A few ideas to make this more aggressive and see if it helps move people off `edge`:

 - make the wording more explicit about what the user needs to do
 - show the dialog every 30mins (or hour) while the app is running, rather than just on launch
 - show the dialog but then close the app when it's dismissed to essentially bomb the app

**Comment by flexiondotorg** [comment]
It seems a more heavy handed approach is indeed required. Anything we can do to help?

**Comment by shiftkey** [comment]
@flexiondotorg not really, just gearing up for another release next week but I guess that's gonna skip Edge 🤷‍♂

**Comment by zypA13510** [comment]
> It's been almost 2 months since I did the cutover to beta and published the dialog about getting people to migrate from `edge` to `beta`. We had almost 400 installs at the time:
> 
> Fast-forward to today and we still have 300+ active installs of `edge`:
> 
> I was really hoping that this number would be much lower, as I'd love to start exploring using `edge` as a preview channel for things like Electron upgrades and experimentation. But I can't do that if users are ignoring the warning and still running the wrong enclosure.

How do you count active installs?
Because coming to this repository today to install it on a new system, I come to realize that I had an old install that hasn't been used for at least 3 months, but the system is regularly maintained. If these were counted then I think it's probably normal?

**Comment by zypA13510** [comment]
> It's been almost 2 months since I did the cutover to beta and published the dialog about getting people to migrate from `edge` to `beta`. We had almost 400 installs at the time:
> 
> Fast-forward to today and we still have 300+ active installs of `edge`:
> 
> I was really hoping that this number would be much lower, as I'd love to start exploring using `edge` as a preview channel for things like Electron upgrades and experimentation. But I can't do that if users are ignoring the warning and still running the wrong enclosure.

How do you count active installs?
Because coming to this repository today to install it on a new system, I come to realize that I had an old install that hasn't been used for at least 3 months, but the system is regularly maintained. If these were counted then I think it's probably normal?

**Comment by shiftkey** [comment]
@zypA13510 

> How do you count active installs?

This is a question for the Snapcraft team. I can't answer this.

**Comment by NatoBoram** [comment]
Hello! I just upgraded to the `beta` channel on a fresh Ubuntu 19.04 install following the prompt and I can't launch it anymore.

```bash
Segmentation fault (core dumped)
```

I think it may explain why some users aren't switching.

**Comment by immackay** [comment]
@NatoBoram, I am under the impression that there is an issue with the 2.1.0 release (snap version 63).

Can you downgrade to version 62? I believe there is no issues with it.

If you don't have version 62 cached, you can download it from the release assets [here](https://github.com/shiftkey/desktop/releases/tag/release-2.0.4-linux1), and install it with
```
snap remove github-desktop
snap install GitHubDesktop-linux-2.0.4-linux1.snap --classic --dangerous
```

See #163 for more information

**Comment by hackel** [comment]
You need to remove (or make optional) the requirement to run this with classic confinement.  As far as I'm concerned, this defeats the whole point of using Snap.  If I didn't want confinement, I would just install the .deb package.  I am not installing some random code from *Microsoft* of all people with unconfined access to my system!  I understand that this currently brings with it several limitations, but that's better than opening up my entire system to random code from the internet.

Eventually I'm sure snapd will have a proper interface for running external editors or whatever the issue is that motivated this change.

**Comment by zypA13510** [comment]
> You need to remove (or make optional) the requirement to run this with classic confinement.

I can think of one way to make this optional - build another snap that's under strict confinement and publish it via a different name, e.g. `github-desktop-confined`. However, I'm not sure this is worth the effort, see my points below.

> As far as I'm concerned, this defeats the whole point of using Snap. If I didn't want confinement, I would just install the .deb package.

Yes, I agree that using snap without confinement is somewhat pointless, save for the auto-update. But, if you didn't want a GitHub Desktop with shell, editor and file explorer integration, what are you using GitHub Desktop for? View git tree? `gitk --all` can do just that (and more, like cherry-picking or tagging). Stage files for commit? Atom or whatever IDE you use will probably have that built-in or provided via a plugin. As far as I'm concerned, GitHub Desktop is more about integrating the workflow, and if those integration features are all missing, this defeats the whole point of using GitHub Desktop.

> I am not installing some random code from _Microsoft_ of all people with unconfined access to my system! I understand that this currently brings with it several limitations, but that's better than opening up my entire system to random code from the internet.

It's not some _random code_, it's code in this repository that you can review or even build yourself. This is open-source, if you so desire, **you can fork it and switch it back to strict confinement, build, install, and use it. No-one will keep you from doing that. The question is, should we build broken snap by default, what do most users expect?**

And it's not just GitHub Desktop, [vscode](https://snapcraft.io/code), [atom](https://snapcraft.io/atom), [webstorm](https://snapcraft.io/webstorm), [sublime](https://snapcraft.io/sublime-text), [android-studio](https://snapcraft.io/android-studio) they all do this, all of them requires classic confinement to function properly.

And if you don't trust the source, the best and only solution is to **not** run it. Would you run an email attachment from an unknown sender, even if in strict snap confinement? Would you run malware in a VM? Who knows it won't utilize some privilege-escalation vulnerability and break out of jail? At some point, you have to trust something.

> Eventually I'm sure snapd will have a proper interface for running external editors or whatever the issue is that motivated this change

That would be great and eventually we can switch back to strict confinement without this discussion.

**Comment by shiftkey** [comment]
Just bumping this issue to indicate the package is unlisted on the Snapcraft store due to various issues and I have no timeline for restoring Snap support. https://github.com/shiftkey/desktop/issues/202 is tracking that work, and I'm blocked on doing any sort of local building or testing of snaps.

**Comment by shiftkey** [comment]
Closing this out as I don't have the bandwidth to revisit this in the short or medium term. If someone else wants to have a shot at this I've added notes in #202 which are from the last time I thought about Snap packaging...
